src/System.Management.Automation/engine/hostifaces/PSTask.cs

```diff
@@ -763,18 +802,28 @@ public bool Add(PSTaskChildJob childJob)
         /// </summary>
         public void StopAll()
         {
+            _stopping = true;
+
             // Accept no more input
             Close();
             _stopAll.Set();
 
             // Stop all running tasks
+            PSTaskBase[] tasksToStop;
             lock (_syncObject)
             {
-                foreach (var task in _taskPool.Values)
-                {
-                    task.Dispose();
-                }
+                tasksToStop = new PSTaskBase[_taskPool.Values.Count];
+                _taskPool.Values.CopyTo(tasksToStop, 0);
+            }
+
+            foreach (var task in tasksToStop)
+            {
+                task.Dispose();
             }
+
+            // Dispose all active runspaces
+            DisposeRunspaces();
+            _stopping = false;
```

Comment:
Is setting to false still needed after stopping everything?

Reply:
Probably not. I just included it because it felt complete.

Reply:
The reason I'm asking is because when looking at the following code in `HandleTaskStateChanged`, I was wondering maybe this code is possible to run even after `StopAll()`. It might be better to remove this line, but it's up to you :)
```
                    if (!_stopping)
                    {
                        // StopAll disposes tasks.
                        task.Dispose();
                    }
```

Reply:
The stop is done synchronously, so I am more worried about tasks that complete before the stop command that might not get disposed. I have done quite a bit of adhoc testing with abrupt stops so I am inclined to will leave things as-is unless there is a compelling reason to change :).  But I really appreciate your input!